src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -56,207 +81,336 @@ def __init__(self,
-        """Returns the `grpc.StatusCode` if the RPC is finished,
-        otherwise first waits until the RPC finishes.
+    async def initial_metadata(self) -> MetadataType:
+        return await self._initial_metadata
 
-        Returns:
-          The `grpc.StatusCode` status code.
-        """
-        if not self.done():
-            try:
-                await self
-            except (asyncio.CancelledError, AioRpcError):
-                pass
+    async def trailing_metadata(self) -> MetadataType:
+        return (await self._status).trailing_metadata()
 
+    async def code(self) -> grpc.StatusCode:
+        await self._status
         return self._code
 
     async def details(self) -> str:
-        """Returns the details if the RPC is finished, otherwise first waits till the
-        RPC finishes.
+        return (await self._status).details()
 
-        Returns:
-          The details.
+    async def debug_error_string(self) -> str:
+        return (await self._status).debug_error_string()
+
+    def _set_initial_metadata(self, metadata: MetadataType) -> None:
+        self._initial_metadata.set_result(metadata)
+
+    def _set_status(self, status: cygrpc.AioRpcStatus) -> None:
+        """Private method to set final status of the RPC.
+
+        This method may be called multiple time due to data race between local
+        cancellation (by application) and C-Core receiving status from peer. We
+        make no promise here which one will win.
         """
-        if not self.done():
-            try:
-                await self
-            except (asyncio.CancelledError, AioRpcError):
-                pass
+        if self._status.done():
+            return
+        else:
+            self._status.set_result(status)
+            self._code = _common.CYGRPC_STATUS_CODE_TO_STATUS_CODE[
+                status.code()]
+
+    async def _raise_rpc_error_if_not_ok(self) -> None:
+        if self._code != grpc.StatusCode.OK:
+            raise _create_rpc_error(await self.initial_metadata(),
+                                    self._status.result())
+
+    def _repr(self) -> str:
+        """Assembles the RPC representation string."""
+        if not self._status.done():
+            return '<{} object>'.format(self.__class__.__name__)
+        if self._code is grpc.StatusCode.OK:
+            return _OK_CALL_REPRESENTATION.format(
+                self.__class__.__name__, self._code,
+                self._status.result().self._status.result().details())
+        else:
+            return _NON_OK_CALL_REPRESENTATION.format(
+                self.__class__.__name__, self._code,
+                self._status.result().details(),
+                self._status.result().debug_error_string())
+
+    def __repr__(self) -> str:
+        return self._repr()
+
+    def __str__(self) -> str:
+        return self._repr()
+
+
+# pylint: disable=abstract-method
+class UnaryUnaryCall(Call, _base_call.UnaryUnaryCall):
```

Comment:
Since the interface `_base_call.UnaryUnaryCall` is a derivated class of `Call` should we narrow the inheritance for the implementation to exclusively the interface `_base_call.UnaryUnaryCall`?

Reply:
The `_base_call.UnaryUnaryCall` is derivated from `_base_call.Call`. Here I hope the concrete implementation of `UnaryUnaryCall` can use logics inside of the concrete implementation of `Call`.